pkg/agent/globalstate.go

```diff
@@ -172,9 +146,170 @@ func (s *agentState) handleVMEventAdded(
+
+	if pod.status.endState != nil {
+		klog.Errorf(
+			"TriggerRestartIfNecessary called with nil endState for pod %v (should only be called after the pod is finished, when endState != nil)",
+			podName,
+		)
+		s.metrics.runnerFatalErrors.Inc()
+		return
+	}
+
+	endTime := pod.status.endState.Time
+
+	if endTime.IsZero() {
+		// If we don't check this, we run the risk of spinning on failures.
+		klog.Errorf("TriggerRestartIfNecessary called with zero'd Time for pod %v", podName)
+		s.metrics.runnerFatalErrors.Inc()
+		// Continue on, but with the time overridden, so we guarantee our minimum wait.
+		endTime = time.Now()
+	}
+
+	// keep this for later.
+	exitKind := pod.status.endState.ExitKind
+
+	switch exitKind {
+	case podStatusExitCanceled:
+		return // successful exit, no need to restart.
+	case podStatusExitPanicked, podStatusExitErrored:
+		// Should restart; continue.
+	default:
+		klog.Errorf("TriggerRestartIfNecessary called with unexpected ExitKind %q for pod %v", podName)
+		s.metrics.runnerFatalErrors.Inc()
+		return
+	}
+
+	// Begin steps (2) and (3) -- wait, then restart.
+	var waitDuration time.Duration
+	totalRuntime := endTime.Sub(pod.status.startTime)
+
+	// If the runner was running for a while, restart immediately.
+	//
+	// NOTE: this will have incorrect behavior when the system clock is behaving weirdly, but that's
+	// mostly ok. It's ok to e.g. restart an extra time at the switchover to daylight saving time.
+	if totalRuntime > time.Second*time.Duration(RunnerRestartMaxWaitSeconds) {
+		waitDuration = 0
+	} else /* Otherwise, randomly pick within RunnerRestartMinWait..RunnerRestartMaxWait */ {
+		r := util.NewTimeRange(time.Second, RunnerRestartMinWaitSeconds, RunnerRestartMaxWaitSeconds)
+		waitDuration = r.Random()
+		klog.Info("Waiting for %s before restarting Runner %v", waitDuration, podName)
+	}
+
+	// Run the waiting (if necessary) and restarting in another goroutine, so we're not blocking the
+	// caller of this function.
+	go func() {
+		if waitDuration != 0 {
+			select {
+			case <-time.After(waitDuration):
+			case <-runnerCtx.Done():
+				klog.Infof("Canceling restart of Runner %v after %s: %s", podName, time.Since(endTime), runnerCtx.Err())
+				return
+			}
+		}
+
+		s.lock.Lock()
+		defer s.lock.Unlock()
+
+		pod, ok := s.pods[podName]
+		if !ok {
+			klog.Warningf("Canceling restart of Runner %v after %s: no longer present in pod map", podName, time.Since(endTime))
+			return
+		}
+
+		pod.status.mu.Lock()
+		defer pod.status.mu.Unlock()
+
+		// Runner was already restarted
+		if pod.status.endState == nil {
+			addedInfo := "this generally shouldn't happen, but could if there's a new pod with the same name"
+			klog.Warningf(
+				"Canceling restart of Runner %v after %s: Runner was already restarted (%s)",
+				podName, time.Since(endTime), addedInfo,
+			)
+		}
+
+		klog.Infof("Restarting %s Runner %v after %s, was running for %s", exitKind, podName, time.Since(endTime), totalRuntime)
```

Comment:
do we still need/want to use exitKind from the calling function? would the value have changed by now?